evalml/pipelines/components/transformers/encoders/onehot_encoder.py

```diff
@@ -137,29 +142,32 @@ def transform(self, X, y=None):
         Returns:
             Transformed dataframe, where each categorical feature has been encoded into numerical columns using one-hot encoding.
         """
-
-        if not isinstance(X, pd.DataFrame):
-            X = pd.DataFrame(X)
-
+        X_copy = _convert_to_woodwork_structure(X)
+        X_copy = _convert_woodwork_types_wrapper(X_copy.to_dataframe())
         cat_cols = self.features_to_encode
 
         if self.parameters['handle_missing'] == "as_category":
-            X[cat_cols] = X[cat_cols].replace(np.nan, "nan")
-        if self.parameters['handle_missing'] == "error" and X.isnull().any().any():
+            for col in cat_cols:
+                if X_copy[col].dtype == 'category' and pd.isna(X_copy[col]).any():
+                    X_copy[col] = X_copy[col].cat.add_categories("nan")
+                    X_copy[col] = X_copy[col].where(pd.isna(X_copy[col]).isin([False]), other='nan')
+
+            X_copy[cat_cols] = X_copy[cat_cols].replace(np.nan, "nan")
+        if self.parameters['handle_missing'] == "error" and X_copy.isnull().any().any():
             raise ValueError("Input contains NaN")
 
         X_t = pd.DataFrame()
         # Add the non-categorical columns, untouched
-        for col in X.columns:
+        for col in X_copy.columns:
             if col not in cat_cols:
-                X_t = pd.concat([X_t, X[col]], axis=1)
+                X_t = pd.concat([X_t, X_copy[col]], axis=1)
         # The call to pd.concat above changes the type of the index so we will manually keep it the same.
         if not X_t.empty:
-            X_t.index = X.index
+            X_t.index = X_copy.index
```

Comment:
Why refer to `X_copy` here instead of `X`?

Reply:
X_copy is the converted-to-woodwork/pandas version; previously, we modified/converted X to a DataFrame so we used X directly. If we use original X now, it could be a numpy object without a `.index` attribute